client/network/src/discovery.rs

```diff
@@ -546,6 +572,22 @@ impl NetworkBehaviour for DiscoveryBehaviour {
 			list.extend(list_to_filter);
 		}
 
+		if list.is_empty() {
+			warn!(target: "sub-libp2p", "Used last-resort-resolutions for {:?}", peer_id);
+
+			let last_resort_resolutions = self
+				.persist_peer_addrs
+				.as_mut()
+				.map(|persist_peer_addrs| {
+					persist_peer_addrs
+						.peer_addrs(peer_id, self.kademlias.values().map(|k| k.protocol_name()))
+				})
+				.into_iter()
+				.flatten();
```

Comment:
I don't think it's a good idea to put this block. It means that the addresses on disk never become obsolete. If you stop your node and restart it a year later, it will use the same addresses as a year ago and they're never purged.

Reply:
Is [that](https://github.com/RGafiyatullin/substrate/blob/78d173ad4ebd6bf7b03412349be81fcd885ce50b/bin/node/cli/tests/peer_persistence_works.rs#L125) the correct illustration of your concern?

Reply:
@tomaka , could [this hack](https://github.com/RGafiyatullin/substrate/commit/0b237ede29e0cfdcb233ccfc81234dd5b447a532) help evicting the stale resolutions? (Given some reasonable default for the age-difference and probably a possibility to tune it via CLI-args for test purposes)

The idea is to prevent the uncontrolled growth of the resolutions cache by:
- limiting the quantity of peer-ids by the LRU cache size;
- keeping only the most fresh resolutions for each peer-id.

Reply:
I was off work for a few days.
After thinking about it, discard my comment.
I prefer to not change the way peers/addresses are pruned, as it could accidentally negatively affect the parachains networking. We could of course change it, but not without a proper future-proof design.

Reply:
IMO I would for now just restore addresses of the sync peerset. The parachains have their own way to fill their peersets.